crates/shared/src/sources/uniswap_v2/pool_fetching.rs

```diff
@@ -25,6 +29,16 @@ pub trait PoolFetching: Send + Sync {
     async fn fetch(&self, token_pairs: HashSet<TokenPair>, at_block: Block) -> Result<Vec<Pool>>;
 }
 
+/// Trait for abstracting the on-chain reading logic for pool state.
+pub trait PoolReading: Send + Sync {
+    fn read_state(
+        &self,
+        pair: TokenPair,
+        batch: &mut Web3CallBatch,
+        block: BlockId,
+    ) -> BoxFuture<'_, Result<Option<Pool>>>;
+}
```

Comment:
Is there a reason why this doesn't use `async_trait`? I think having that consistent across the code base would be nice.

Reply:
Yes there is. Specifically, when making a "batched" fetching interface, you need to add Ehereum RPC request to the batch synchronously and then return a future that combines the result of those batched RPC requests into a `Result<Pool>`. If you use a `async_trait`, then the requests will only get added to the batch after the first poll.

I will add a comment explaining this.